kubevirt_api/tests/test_unit.py

```diff
@@ -0,0 +1,253 @@
+# (C) Datadog, Inc. 2024-present
+# All rights reserved
+# Licensed under a 3-clause BSD style license (see LICENSE)
+
+from copy import deepcopy
+from unittest.mock import MagicMock  # noqa: F401
+
+import pytest
+
+from datadog_checks.dev.utils import get_metadata_metrics
+from datadog_checks.kubevirt_api import KubeVirtApiCheck
+
+from .conftest import mock_http_responses
+from .mock_response import GET_VMIS_RESPONSE, GET_VMS_RESPONSE
+
+pytestmark = [pytest.mark.unit]
+
+healthz_tags = [
+    "endpoint:https://10.244.0.38:443/healthz",
+    "pod_name:virt-api-98cf864cc-zkgcd",
+    "kube_namespace:kubevirt",
+    "kube_cluster_name:test-cluster",
+]
+
+
+def test_check_collects_all_metrics(dd_run_check, aggregator, instance, mocker):
+    mocker.patch("requests.get", wraps=mock_http_responses)
+
+    check = KubeVirtApiCheck("kubevirt_api", {}, [instance])
+
+    check._setup_kube_client = lambda: None
+    check.kube_client = MagicMock()
+    check.kube_client.get_vms.return_value = GET_VMS_RESPONSE["items"]
```

Comment:
Can we mock the client dependency rather than the integration client code? So instead of mocking `check.kube_client`, mock `check.kube_client.custom_obj_client`? This will help ensure there's coverage for all the `kube_client` code. It also might be nice to create one `MockKubeClient` that every test can use so we don't need to manually set the return values in every test.

Reply:
I decided to mock the whole kube_client because I only care about the interface we use (get_vms, get_vmis, etc...) instead of the implementation, I didn't want the tests to rely on the implementations.
And we have different return values for different tests, that's why I didn't refactor it yet.
What do you think?

Reply:
I see, but by doing this we are not testing all of our code. I think it's important to have good coverage and only mock the external inputs (in this case the kubeclient dependency).

Reply:
My 2¢:

The mocking level is appropriate. [Mock what you own](https://hynek.me/articles/what-to-mock-in-5-mins/). That being said the point about coverage is well taken, so the client needs to be as thin as possible. That's what we have here, almost no logic other than calls to the `kubernetes` lib.

I have some improvement suggestions:

1. use [autospecing](https://docs.python.org/3/library/unittest.mock.html#auto-speccing), it's a safer way to mock in such a dynamic language like Python.
2. Don't reach for private methods/attrs in your tests. Imagine that tests are your client code. So let's just own that we have a public dependency on `kube_client`. From here you have options. You could turn `kube_client` into a `property` with a setter. You put logic you have rn in `_setup_kube_client` into the property getter, then use the setter to let your unit tests set the property to a mock. Alternatively you could make `_setup_kube_client` public and have it return the client (which you then assign to a local variable). In that case you mock the return value of this `_setup_kube_client`.